test/dynamo/test_export.py

```diff
@@ -2190,6 +2183,59 @@ def forward(self, *args):
         res = gm(input_tensor, input_tensor2)
         self.assertTrue(torch._dynamo.utils.same(ref, res))
 
+    def test_export_mark_dynamic_conflict_dynamic_dim(self):
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] > 3:
+                return x.sin()
+            return x.cos()
+
+        torch._dynamo.mark_dynamic(y, 0)
+        with self.assertRaisesRegex(
+            AssertionError, "Illegal to export tensor already marked dynamic"
+        ):
+            torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
+
+    def test_export_dynamic_dim_duplicate_dims(self):
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] > 3:
+                return x.sin()
+            return x.cos()
+
+        with self.assertRaisesRegex(AssertionError, "Duplicate index 0"):
+            torch._dynamo.export(
+                my_dyn_fn, y, constraints=[dynamic_dim(y, 0), dynamic_dim(y, 0)]
+            )
+
+    def test_export_dynamic_dim_unknown_tensor(self):
+        y = torch.randn([3, 3, 3])
+        z = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] > 3:
+                return x.sin()
+            return x.cos()
+
+        with self.assertRaisesRegex(
+            AssertionError, "User specified dynamic dim for unknown tensors"
+        ):
+            torch._dynamo.export(
+                my_dyn_fn, y, constraints=[dynamic_dim(z, 0), dynamic_dim(y, 0)]
+            )
+
+    @config.patch(dynamic_shapes=True)
+    def test_export_dynamic_dim_cleanup(self):
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            return x.cos()
+
```

Comment:
Nit:
Construct dynamic_dim before `export` call and asserts that dynamic_dim doesn't add state to `y` either.

Reply:
good call